lib/search_files.js

```diff
@@ -0,0 +1,291 @@
+var fs = require('fs');
+var os = require('os');
+var path = require('path');
+var glob = require('glob');
+
+
+function isWindows() {
+  return process.platform === 'win32';
+}
+
+function isFunction(v) {
+  return (typeof v === 'function');
+}
+
+function isObject(v) {
+  return ({}.toString.call(v) === '[object Object]');
+}
+
+function isArray(v) {
+  return Array.isArray(v);
+}
+
+function isString(v) {
+  return (typeof v === 'string');
+}
+
+function doNothing(v) {
+  return v;
+}
+
+
+function firstUnNullElem(arr, fn /*, ...*/) {
+  var args = Array.prototype.slice.call(arguments, 2);
+
+  for (var i = 0, n = arr.length; i < n; i++) {
+    var ret = fn.apply(arr, [arr[i]].concat(args));
+    if (ret) {
+      return ret;
+    }
+  }
+  return null;
+}
+
+function eachUnNullElem(arr, fn /*, ...*/) {
+  var args = Array.prototype.slice.call(arguments, 2);
+
+  var ret = [];
+  for (var i = 0, n = arr.length; i < n; i++) {
+    var v = fn.apply(arr, [arr[i]].concat(args));
+    if (v) {
+      if (isArray(v)) {
+        Array.prototype.push.apply(ret, v);
+      } else {
+        ret.push(v);
+      }
+    }
+  }
+  return ret;
+}
+
+function eachProp(obj, fn /*, ...*/) {
+  var args = Array.prototype.slice.call(arguments, 2);
+
+  var ret = {};
+  var keys = Object.keys(obj);
+  for (var i = 0, n = keys.length; i < n; i++) {
+    var key = keys[i];
+    ret[key] = fn.apply(obj, [obj[key], key].concat(args));
+  }
+  return ret;
+}
+
+
+function parsePath(pth) {
+  return path.parse ? path.parse(pth) : require('path-parse')(pth);
+}
+
+function isAbsolutePath(pth) {
+  return path.isAbsolute ? path.isAbsolute(pth) : !!parsePath(pth).root;
+}
+
+function normalizeHomedir(pth, homedir) {
+  if (!pth) {
+    return pth;
+  }
+
+  if (pth[0] === '~') {
+    if (pth.length === 1) {
+      return homedir;
+    }
+
+    if (pth[1] === '/') {
+      return path.join(homedir, pth.slice(2));
+    }
+
+    if (isWindows() && pth[1] === '\\') {
+      return path.join(homedir, pth.slice(2));
+    }
+  }
+
+  return pth;
+}
+
+function normalizePath(pth, ctx) {
+  if (isFunction(pth)) {
+    pth = pth();
+  }
+
+  pth = normalizeHomedir(pth, ctx.homedir);
+
+  if (!isAbsolutePath(pth)) {
+    return path.join(ctx.cwd, pth);
+  }
+
+  return path.normalize(pth);
+}
+
+
+var initCtx = {
+  homedir: os.homedir ? os.homedir() : require('userhome')(),
+  cwd: process.cwd(),
+};
+
+function normalizeContext(ctx) {
+  ctx = eachProp(ctx || {}, doNothing);
+
+  ctx.cwd = !ctx.cwd ? initCtx.cwd :
+    normalizePath(ctx.cwd, initCtx);
+
+  ctx.homedir = !ctx.homedir ? initCtx.homedir :
+    normalizePath(ctx.homedir, initCtx);
+
+  return ctx;
+}
+
+function normalizeExtensions(exts) {
+  if (isFunction(exts)) {
+    exts = exts();
+  }
+
+  if (isString(exts)) {
+    return [exts];
+  }
+
+  if (isArray(exts)) {
+    return exts.filter(isString);
+  }
+
+  if (isObject(exts)) {
+    return Object.keys(exts).sort();
```

Comment:
This part is for getting a property array of `exts` object and each properties is unique. Since being unstable of `.sort` is about comparing equal, there should be no problem in this case.